### ISSUE TITLE
[v2 FEATURE]: An option to input a test set separately and use multiple `--num-folds` only for train/val split

### ISSUE THREAD

**Issue by yunsiechung** [chemprop/chemprop#1080]
I'm a senior scientist at Merck, which is part of the MLPDS Consortium. We would like a new feature to only specify a test set and let Chemprop randomly split train/val with multiple split folds.

**Is your feature request related to a problem? Please describe.**
With Chemprop v1, we provide a test set separately and then have Chemprop randomly split train/val with multiple `num_folds`. We usually generate a total of 4 submodels, with two different train/val split seeds (`num_folds` each with different `seed`) and with two ensemble seeds for each train/val split (`ensemble_size` each with different `pytorch_seed`). Our training CLI scripts look like this for version 1:
```shell
chemprop_train \\
--dataset_type regression \\
--data_path train_val.csv \\
--separate_test_path test.csv \\
--split_sizes 0.90 0.10 0.0 \\
--num_folds 2 \\
--ensemble_size 2
``` 
This allowed Chemprop use the same test set (`test.csv`) for all 4 submodels while it randomly splits`train_val.csv` into 90% training and 10% validation sets with two different seeds. We gives a bit better performance than just using 4 ensembles.

However, with Chemprop v2, this cannot be done easily. We have to manually split train/val and train two separate models with different train/val/test splits. We would like Chemprop v2 to be able to take a test set as a separate input argument and splits train/val with multiple data split folds while keeping the test set same.

**Desired solution/workflow**
It would be great if `--separate-test-path` argument becomes available in v2 for model training.

**Comment by KnathanM** [comment]
Hi Yunsie,

Based on #1073, I assume you have seen that you can combine your train/val data with your test data into one csv and then use a separate splits json file to specify the splits. You are right that this puts extra work on users. We originally decided to not support `--separate-test-path` because then we could also need to copy all the other dataset arguments for that separate test file. In v1 this included:
`--separate-test-descriptors-path`
`--separate-test-atom-features-path`
`--separate-test-atom-descriptors-path`
`--separate-test-bond-features-path`
`--separate-test-constraints-path`
`--separate-test-phase-features-path`
It could also include:
`--separate-test-no-header-row`
`--separate-test-smiles-columns`
`--separate-test-reaction-columns`
`--separate-test-target-columns`
`--separate-test-ignore-columns`

In v1 they also had a separate val test path flag which also requires all these flags to be added for the val data. We felt it was much simpler have only one SMILES+targets file, one atom features file, one bond features file, etc. and have a splits file that acts on each of those data files. 

How difficult is it to have a separate script split your train/val data and then combine it with the test data to get it in the format expected by the current chemprop? If it is difficult, I would consider adding the `--separate-test-path` flag, especially because having a separate test file is common in ML workflows. I opened a PR on my fork that shows roughly how I would do this. https://github.com/KnathanM/chemprop/pull/4/files

**Comment by KnathanM** [comment]
On that branch on my fork, I tested using this command:
`chemprop train -i tests/data/regression/mol/mol.csv --split-sizes 0.9 0.1 0 --num-replicates 2 --ensemble-size 2 --separate-test-path tests/data/regression/mol/mol.csv `

**Comment by yunsiechung** [comment]
Hi Shih-Cheng, thanks for your explanation. It's not the biggest effort for us to split train/val on our own. It's just that with the current version, we have to separately train the model twice with each train/val/test split we want to use, which is also not the biggest effort. We just miss having an option to only use `-num-folds` for train/val split.

Would it be then possible for the user to only specify `test` split in a separate split json file? And have Chemprop split train/val on their own with the given `-num-folds` and split ratio whenever only `test` is specified in the input split file. 

Also, the current documentation on how to format the `--splits-file` is poor ([link](https://chemprop.readthedocs.io/en/main/tutorial/cli/train.html))
![image](https://github.com/user-attachments/assets/a1ae0ec0-7f85-427f-97b4-b6a4bad32d6a)
The sample command shows that it's a csv file but it doesn't look like a csv file. Moreover, it's unclear whether the indexing starts from `0` or `1`. Normally, I would think that the indexing starts from `0` but it seems like it starts from `1`. Could the documentation for `--splits-file` be improved? Can multiple train/val/test splits be specified in the splits file?

I would appreciate your help.

**Comment by KnathanM** [comment]
You are right about it not being a CSV file, it is a JSON file. We just fixed that yesterday at 3. I also agree it is unclear if the indexing starts at 0 or 1. It starts at 0. I'll update that documentation. Is this more clear:

"""
`chemprop train --splits-file splits.json -i data.csv -t regression`
splits.json
```
    [
        {"train": [0, 1, 2], "val": "3-5", "test": "6,7"},
        {"val": [0, 1, 2], "test": "3-5", "train": "6,7"},
    ]
```
The splits.json file contains a list of dictionaries which has keys: `train`, `val`, and `test` and values: indices. The indices can be either a list (e.g. `[0,1,3]`) or a string (e.g. `"12,28, 2"`). If the indices is a string, ranges can be specified with a dash (e.g. `"3-5"`). Note that ranges are inclusive. For example, the first 100 data points would be specified using `"0-99"`.  

If there are multiple dictionaries in the list, then multiple models will be trained. For example, 3-fold cross validation could look like: 
```
    [
        {"train": [0, 1], "val": [2, 3], "test": [4, 5]},
        {"val": [0, 1], "test": [2, 3], "train": [4, 5]},
        {"test": [0, 1], "train": [2, 3], "val": [4, 5]},
    ]
```
"""

**Comment by KnathanM** [comment]
Your suggestions are good and I'll bring them up at our next developer meeting.

Until then, I think you can still do what you'd like with the current chemprop version**. This is what I envision:
Make a script called `cv_with_separate_test.py`
```
from argparse import ArgumentParser
import pandas as pd
import json
from sklearn.model_selection import KFold


def main():
    parser = ArgumentParser(
        description="Combine train/val data with test data and create a corresponding splits.json "
        "file using cross-validation on the train/val data."
    )
    parser.add_argument("--data_path", required=True, help="Path to the train/val data CSV file.")
    parser.add_argument(
        "--separate_test_path", required=True, help="Path to the separate test data CSV file."
    )
    parser.add_argument(
        "--num_folds", type=int, required=True, help="Number of folds for cross-validation."
    )
    args = parser.parse_args()

    data_df = pd.read_csv(args.data_path)
    separate_test_df = pd.read_csv(args.separate_test_path)
    combined_df = pd.concat([data_df, separate_test_df], ignore_index=True)
    combined_df.to_csv("train_val_test.csv", index=False)

    num_lines_data = len(data_df)
    num_lines_combined = len(combined_df)
    test_range = f"{num_lines_data}-{num_lines_combined - 1}"

    kf = KFold(n_splits=args.num_folds)
    x = list(range(num_lines_data))
    splits = []
    for train_idx, val_idx in kf.split(x):
        splits.append({"train": train_idx.tolist(), "val": val_idx.tolist(), "test": test_range})

    with open("splits.json", "w") as f:
        json.dump(splits, f, indent=4)


if __name__ == "__main__":
    main()
```
Call this script using `python cv_with_separate_test.py --data_path <path> --separate_test_path <path> --num_folds 2` and then call chemprop using `chemprop train -i train_val_test.csv --splits-file splits.json --ensemble-size 2`.

Note that this script doesn't use `splits_sizes` because that is mutually exclusive with `num_folds`. Best I can tell though `num_folds` is only used in v1 if you also specify `--split_type cv`. And in any event, if looks like you want to do two replicates of a random 90/10 split on your train/val data. So here is a version that does this:
```
from argparse import ArgumentParser
import pandas as pd
import json
from sklearn.model_selection import ShuffleSplit


def main():
    parser = ArgumentParser(
        description="Combine train/val data with test data and create a corresponding splits.json "
        "file using cross-validation on the train/val data."
    )
    parser.add_argument("--data_path", required=True, help="Path to the train/val data CSV file.")
    parser.add_argument(
        "--separate_test_path", required=True, help="Path to the separate test data CSV file."
    )
    parser.add_argument(
        "--num_replicates", type=int, required=True, help="Number of times to re-split data."
    )
    parser.add_argument(
        "--split_sizes",
        type=float,
        nargs=3,
        required=True,
    )
    args = parser.parse_args()

    data_df = pd.read_csv(args.data_path)
    separate_test_df = pd.read_csv(args.separate_test_path)
    combined_df = pd.concat([data_df, separate_test_df], ignore_index=True)
    combined_df.to_csv("train_val_test.csv", index=False)

    num_lines_data = len(data_df)
    num_lines_combined = len(combined_df)
    test_range = f"{num_lines_data}-{num_lines_combined - 1}"

    ss = ShuffleSplit(n_splits=args.num_replicates, test_size=args.split_sizes[1])
    x = list(range(num_lines_data))
    splits = []
    for train_idx, val_idx in ss.split(x):
        splits.append({"train": train_idx.tolist(), "val": val_idx.tolist(), "test": test_range})

    with open("splits.json", "w") as f:
        json.dump(splits, f, indent=4)


if __name__ == "__main__":
    main()
```
`python cv_with_separate_test.py --data_path <path> --separate_test_path <path> --num_replicates 2 --split_sizes 0.9 0.1 0`


** current intended version. While working on this I found a bug that is fixed in #1086.

**Comment by JacksonBurns** [comment]
@yunsiechung we had some offline discussion about this feature - what do you think about us allowing `--data-path` and the associated feature arguments to accept _multiple_ values (i.e. `--data-path train_and_val.csv test.csv`) such that Chemprop could then automatically do splitting on the training and validation data while using the second file as the testing data?

**Comment by yunsiechung** [comment]
@KnathanM Thanks for explaining how `--splits-file` works. I didn't know it could take multiple data splits there. That's really helpful.
It seems like `--num_replicates` in v2 is equivalent/similar to `--num_folds` in v1, is that correct? I don't know why it's called `--num_replicates` because replicates sound like you are generating the exact same copy of the data split. I'm just curious whether it is a commonly used term. Wouldn't a different name like `--num_data_splits` be more straight-forward? I was initially very confused with the name `--num_replicates`. Correct me if I am misunderstanding anything.

@JacksonBurns Your suggestion would work great! Then for additional descriptors (e.g. atom, bond, molecular) we are passing in, do we just pass the descriptors relevant to test.csv as a second file?

Thank you so much for your help!! This will make our model training easier.

**Comment by JacksonBurns** [comment]
> That's really helpful. It seems like `--num_replicates` in v2 is equivalent/similar to `--num_folds` in v1, is that correct? I don't know why it's called `--num_replicates` because replicates sound like you are generating the exact same copy of the data split. I'm just curious whether it is a commonly used term. Wouldn't a different name like `--num_data_splits` be more straight-forward? I was initially very confused with the name `--num_replicates`. Correct me if I am misunderstanding anything.

Cross Validation is a form of pseudo-replication, whereas resampling is true replication. Therefore the former gets a special name ("folds") and the latter gets the 'real' name ("replicates").

> Then for additional descriptors (e.g. atom, bond, molecular) we are passing in, do we just pass the descriptors relevant to test.csv as a second file?

Yes, it would look something like `chemprop train --data-path train.csv val.csv test.csv --atom-features-path train_atom_f.npz val_atom_f.npz test_atom_f.npz` if one were specifying all three sets, for example.

**Comment by KnathanM** [comment]
Re: calling the flag num_replicates, here we are drawing an analogy to [replication](https://en.m.wikipedia.org/wiki/Replication_(statistics)) in experiment design. If you repeat an experiment without changing any of the parameters, the outcome may differ slightly due to random chance. It is advisable to do multiple repeats (replicates) to get an idea of the true outcome of the experiment. 

With cross validation, you don’t expect the data distribution to be different between folds, but it is more likely than a true random re-sampling from the data set. 

I also thought about calling it num_data_splits or num_splitting_schemes, but in the end landed on replicates. I will add an explanation to the docs because I expect you won't be the only one confused by this.

**Comment by yunsiechung** [comment]
I still think num_replicates is a very confusing name, but it is your call. Thanks for the explanation.
